## Lab5
Task
Utilize the diabetes dataset from lab 4. Perform cross-validation on nine polynomial models, ranging from degree 0 to 8. (2 points)
Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. Calculate the mean value and standard deviation of these metrics from the cross-validation. Include both values. (2 points)
Identification of the Best Model: Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model. (2 points)

In [3]:
from sklearn import datasets
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import pandas as pd


In [2]:
# Read dataset
X, y = datasets.load_diabetes(as_frame=True, scaled=False, return_X_y=True)


In [7]:
# Cross validation
cv_result=[]
for degree in range(0,9):
    
    model = make_pipeline(PolynomialFeatures(degree), LinearRegression())
    
    r2_cv = cross_val_score(model, X, y, cv=5, scoring='r2')
    r2_mean = r2_cv.mean()
    r2_std = r2_cv.std()

    mae_cv = -cross_val_score(model, X, y, cv=5, scoring='neg_mean_absolute_error')
    mae_mean = mae_cv.mean()
    mae_std = mae_cv.std()

    cv_result.append([degree,r2_mean, r2_std, mae_mean, mae_std])

# Display result    
cv_df = pd.DataFrame(cv_result, columns=['DEGREE','R2_MEAN','R2_STD', 'MAE_MEAN', 'MAE_STD'])
print(cv_df)


   DEGREE       R2_MEAN        R2_STD     MAE_MEAN     MAE_STD
0       0     -0.027506      0.036772    66.045624    3.474660
1       1      0.482316      0.049269    44.276499    2.100110
2       2     -0.528281      1.086913    66.857204   24.676461
3       3   -203.420273    225.879811   342.052529  142.438254
4       4   -571.083108    369.891883   657.260477  159.475902
5       5   -436.856887    379.100423   562.993636   59.917202
6       6  -1694.699196   2631.163241   742.502338  191.024419
7       7  -5530.894075   9518.586955  1032.681725  393.439617
8       8 -16076.255118  28049.952581  1475.658532  706.280274


In [8]:

# Find best model
cv_df_sorted = cv_df.sort_values(by=['R2_MEAN', 'MAE_MEAN'], ascending=[False, True])
best_fit_model = cv_df_sorted.iloc[0]
best_fit_model

DEGREE       1.000000
R2_MEAN      0.482316
R2_STD       0.049269
MAE_MEAN    44.276499
MAE_STD      2.100110
Name: 1, dtype: float64

From the result, we can see that the model with degree 1 is the best model. It has highest R squared mean value(0.482316) and smallest MAE mean(44.2764) value .